### PR TITLE
gh-111489: Add PyList_FromArrayMoveRef() function

### PR DESCRIPTION
**Pull request by vstinner**
Add new functions to create tuple and list from arrays:

* PyTuple_FromArray().
* PyTuple_FromArrayMoveRef().
* PyList_FromArrayMoveRef().

Add tests on new functions.

* Issue: gh-111489

----
:books: Documentation preview :books:: https://cpython-previews--112975.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
I tried to clear array items (or set bytes to a pattern, such as PYMEM_DEADBYTE=`0xDD`) in `PyTuple_FromArrayMoveRef()` and `PyList_FromArrayMoveRef()` in debug mode, but it broke current code which expect the array to remain usable. Pseudo-code:

```c
PyObject *const * args = &PyTuple_GET_ITEM(callargs, 0);
for (size_t i = 0; i < argcount; i++) {
    Py_INCREF(args[i]);
}

PyObject *u = _PyTuple_FromArraySteal(args, argcount);
// ... check if u == NULL ...

// ... code using callargs ...
```

So I chose to leave *array* unchanged.

**Comment by sweeneyde** [comment]
I think it's confusing that the function consumes references when it succeeds but does not consume them when it fails. I would think that it should always consume references or never consume references.

If it always consumes references, then one can write

```
    PyObject *pair[2];
    pair[0] = pair[1] = something;
    return _PyList_FromArraySteal(pair);
```

I think the only-sometimes-consuming was why there used to be a lot of error-handling mistakes in module initialization code with `PyModule_AddObject`. The [docs](api/module.html#c.PyModule_AddObject) note that

> Unlike other functions that steal references, PyModule_AddObject() only releases the reference to value on success.

I would think that we would want to replicate the always-consuming "other functions" (e.g. PyTuple_SetItem) style, not the only-sometimes-consuming PyModule_AddObject style.

Is there a reason to prefer this style?

File: Doc/c-api/tuple.rst [hunk 1]
```diff
@@ -36,6 +36,30 @@ Tuple Objects
    Return a new tuple object of size *len*, or ``NULL`` on failure.
 
 
+.. c:function:: PyObject* PyTuple_FromArray(PyObject *const *array, Py_ssize_t size)
+
+   Create a tuple of *size* items and copy references from *array* to the new
+   tuple.
+
+   * Return a new reference on success.
+   * Set an exception and return NULL on error.
```
Comment (serhiy-storchaka) [inline Doc/c-api/tuple.rst:45]:
```suggestion
   * Set an exception and return ``NULL`` on error.
```

File: Lib/test/test_capi/test_list.py [hunk 2]
```diff
@@ -342,3 +342,11 @@ def test_list_extend(self):
 
         # CRASHES list_extend(NULL, [])
         # CRASHES list_extend([], NULL)
+
+    def test_list_fromarraymoveref(self):
+        # Test PyList_FromArrayMoveRef()
+        list_fromarraymoveref = _testcapi.list_fromarraymoveref
+
+        lst = [object() for _ in range(5)]
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_list.py:350]: I would use the following example:
```suggestion
        lst = [[i] for _ in range(5)]
```
Better repr.

File: Lib/test/test_capi/test_list.py [hunk 3]
```diff
@@ -342,3 +342,11 @@ def test_list_extend(self):
...
+        copy = list_fromarraymoveref(lst)
+        self.assertEqual(copy, lst)
```
Comment (serhiy-storchaka) [inline Lib/test/test_capi/test_list.py:352]: Test also with zero and negative size. with too large size (causing MemoryError or OverflowError), with array=NULL, and with NULLs in the array. We should know what cases raise exceptions, what cases crash and what cases are valid, and how errors are handled.

File: Objects/tupleobject.c [hunk 4]
```diff
@@ -371,47 +371,60 @@ tupleitem(PyTupleObject *a, Py_ssize_t i)
...
-    PyTupleObject *tuple = tuple_alloc(n);
+    PyTupleObject *tuple = tuple_alloc(size);
     if (tuple == NULL) {
-        for (Py_ssize_t i = 0; i < n; i++) {
-            Py_DECREF(src[i]);
+        for (Py_ssize_t i = 0; i < size; i++) {
+            Py_DECREF(array[i]);
         }
         return NULL;
     }
```
Comment (serhiy-storchaka) [inline Objects/tupleobject.c:406]: It does not match the documentation.

**Review by serhiy-storchaka — COMMENTED** [review]
The Ref suffix is used in many new functions, would not it cause confusion?

What if use `PyTuple_FromArrayMove()` or `PyTuple_FromArrayMoveOwnership()`? Or maybe `PyTuple_CopyFromArray()` and `PyTuple_MoveFromArray()`?

Why there is no copying API for list?

**Comment by vstinner** [comment]
> What if use PyTuple_FromArrayMove() or PyTuple_FromArrayMoveOwnership()? Or maybe PyTuple_CopyFromArray() and PyTuple_MoveFromArray()?

I suggest discussing the name at: https://github.com/capi-workgroup/api-evolution/issues/25

File: Lib/test/test_capi/test_list.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/test_capi/test_list.py:350]: First, I used `list("abdef")`, but all items are immortal and so tests didn't detect refcount bugs in my C code. I used `object()` to make sure that each object is unique.

I don't think that repr is important here, the test is not supposed to fail :-)

**Comment by vstinner** [comment]
> Why there is no copying API for list?

I only exposed what already existed. But yeah, when I wrote the PR, I also noticed that it seems like there is "a hole" in the API :-) I can add a "copy" API for list as well.

**Comment by vstinner** [comment]
I convert this PR to a draft until we agree on naming: https://github.com/capi-workgroup/api-evolution/issues/25

**Comment by vstinner** [comment]
> I convert this PR to a draft until we agree on naming: https://github.com/capi-workgroup/api-evolution/issues/25

I created this PR to have a more concrete API to discuss the naming.

I prefer to close the PR for now, until we agree on naming.

**Pull request closed without merging**